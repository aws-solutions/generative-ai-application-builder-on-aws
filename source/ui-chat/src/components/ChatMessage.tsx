/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { FC, memo, useContext, useState } from 'react';
import { Button, Icon } from '@cloudscape-design/components';
import HomeContext from '../home/home.context';
import { MessageWithSource } from '../types/chat';
import { MemoizedReactMarkdown } from './MemoizedReactMarkdown';
import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';
import rehypeExternalLinks from 'rehype-external-links';
import { CodeBlock } from './CodeBlock';
import { Components } from 'react-markdown';

import { SourceDocumentsModal } from './SourceDocumentsModal';

export interface Props {
    message: MessageWithSource;
    messageIndex: number;
    displaySourceConfigFlag: boolean;
}

const MARKDOWN_COMPONENTS: Components = {
    code: ({ node, className, children, ...props }) => {
        const match = /language-(\w+)/.exec(className || '');

        return match ? (
            <CodeBlock language={(match && match[1]) || ''} value={String(children).replace(/\n$/, '')} {...props} />
        ) : (
            <code className={className} {...props}>
                {children}
            </code>
        );
    },
    table: ({ children }) => {
        return <table>{children}</table>;
    },
    th: ({ children }) => {
        return <th className="break-words border border-black bg-gray-500 px-3 py-1 text-white">{children}</th>;
    },
    td: ({ children }) => {
        return <td className="break-words border border-black px-3 py-1">{children}</td>;
    }
};

export const ChatMessage: FC<Props> = memo(({ message, messageIndex, displaySourceConfigFlag }) => {
    const {
        state: { selectedConversation, messageIsStreaming }
    } = useContext(HomeContext);

    const [messagedCopied, setMessageCopied] = useState(false);
    const [docSourceModalVisible, setDocSourceModalVisible] = useState(false);
    const onModalDismiss = () => setDocSourceModalVisible(false);

    const copyOnClick = () => {
        if (!navigator.clipboard) return;

        navigator.clipboard.writeText(message.content).then(() => {
            setMessageCopied(true);
            setTimeout(() => {
                setMessageCopied(false);
            }, 2000);
        });
    };

    const displaySourceInResponse =
        displaySourceConfigFlag && message.sourceDocuments && message.sourceDocuments.length > 0;

    return (
        <div className={`${message.role === 'assistant' ? 'bg-gray-100' : ''}`} style={{ overflowWrap: 'anywhere' }}>
            <div className=" m-auto flex md:gap-6 md:py-6 xl:max-w-2xl">
                {message.role === 'assistant' && (
                    <div className="text-right font-bold ">
                        <Icon name={'contact'} />
                    </div>
                )}

                <div className="prose mt-[-2px] w-full">
                    {message.role === 'user' ? (
                        <div className="w-full">
                            <MemoizedReactMarkdown
                                remarkPlugins={[remarkGfm, remarkMath]}
                                rehypePlugins={[
                                    [rehypeExternalLinks, { target: '_blank', rel: 'noopener noreferrer' }]
                                ]}
                                unwrapDisallowed
                                components={MARKDOWN_COMPONENTS}
                            >
                                {message.content}
                            </MemoizedReactMarkdown>
                        </div>
                    ) : (
                        <div className="flex ">
                            <div className="flex-1">
                                <MemoizedReactMarkdown
                                    className="prose dark:prose-invert flex-1"
                                    remarkPlugins={[remarkGfm, remarkMath]}
                                    rehypePlugins={[
                                        [rehypeExternalLinks, { target: '_blank', rel: 'noopener noreferrer' }]
                                    ]}
                                    unwrapDisallowed
                                    components={MARKDOWN_COMPONENTS}
                                >
                                    {`${message.content}${
                                        messageIsStreaming &&
                                        messageIndex === (selectedConversation?.messages.length ?? 0) - 1
                                            ? '‚ñç'
                                            : ''
                                    }`}
                                </MemoizedReactMarkdown>
                            </div>
                            <div className="copy-button-div md:-mr-8">
                                {messagedCopied ? (
                                    <Icon name="status-positive" />
                                ) : (
                                    <div className="copy-button-div-nested">
                                        <Button
                                            data-testid="copy-msg-button"
                                            iconName="copy"
                                            variant="icon"
                                            onClick={copyOnClick}
                                        />
                                    </div>
                                )}
                                {displaySourceInResponse && (
                                    <div>
                                        <Button
                                            data-testid="source-docs-button"
                                            iconName="status-info"
                                            variant="icon"
                                            onClick={() => setDocSourceModalVisible(true)}
                                        />

                                        <SourceDocumentsModal
                                            visible={docSourceModalVisible}
                                            onDismiss={onModalDismiss}
                                            sourceDocumentsData={message.sourceDocuments}
                                        />
                                    </div>
                                )}
                            </div>
                        </div>
                    )}
                </div>
                {message.role === 'user' && (
                    <div className="min-w-[40px] text-right">
                        <Icon name={'user-profile'} />
                    </div>
                )}
            </div>
        </div>
    );
});
ChatMessage.displayName = 'ChatMessage';
