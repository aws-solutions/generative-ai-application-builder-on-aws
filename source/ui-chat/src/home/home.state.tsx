/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Conversation } from '../types/chat';
import { MAX_PROMPT_TEMPLATE_LENGTH, MODEL_MAX_INPUT_LENGTH } from '../utils/constants';

export interface UseCaseConfigType {
    KnowledgeBaseParams: {
        ReturnSourceDocs: boolean;
        KnowledgeBaseType: string;
        KendraKnowledgeBaseParams: {
            ExistingKendraIndexId: string;
            RoleBasedAccessControlEnabled: boolean;
        };
        NumberOfDocs: number;
        ScoreThreshold: number;
        NoDocsFoundResponse: string;
    };
    ConversationMemoryParams: {
        HumanPrefix: string;
        ConversationMemoryType: string;
        ChatHistoryLength: number;
        AiPrefix: string;
    };
    UseCaseName: string;
    LlmParams: {
        Streaming: boolean;
        Temperature: number;
        Verbose: boolean;
        BedrockLlmParams: {
            ModelId: string;
        };
        ModelProvider: string;
        PromptParams: {
            UserPromptEditingEnabled: boolean;
            DisambiguationEnabled: boolean;
            MaxInputTextLength: number;
            RephraseQuestion: boolean;
            PromptTemplate: string;
            MaxPromptTemplateLength: number;
            DisambiguationPromptTemplate: string;
        };
        ModelParams: object;
        RAGEnabled: boolean;
    };
}

export interface HomeInitialState {
    loading: boolean;
    messageIsStreaming: boolean;
    selectedConversation: Conversation | undefined;
    promptTemplate: string;
    defaultPromptTemplate: string;
    RAGEnabled: boolean;
    useCaseConfig: UseCaseConfigType;
    userPromptEditingEnabled: boolean;
    maxPromptTemplateLength: number;
    maxInputTextLength: number;
}

export const initialState: HomeInitialState = {
    loading: false,
    messageIsStreaming: false,
    selectedConversation: {
        id: '',
        name: 'New Conversation',
        messages: []
    },
    promptTemplate: '',
    defaultPromptTemplate: '',
    RAGEnabled: false,
    useCaseConfig: {} as UseCaseConfigType,
    userPromptEditingEnabled: true,
    maxPromptTemplateLength: MAX_PROMPT_TEMPLATE_LENGTH,
    maxInputTextLength: MODEL_MAX_INPUT_LENGTH
};
