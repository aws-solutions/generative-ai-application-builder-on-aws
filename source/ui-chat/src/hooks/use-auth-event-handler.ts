// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { useCallback, useEffect } from 'react';
import { fetchAuthSession, Hub } from '@aws-amplify/core';
import { signOut, signInWithRedirect } from '@aws-amplify/auth';
import { useSelector } from 'react-redux';
import { RootState } from '../store/store';

/**
 * Interface defining the return type of the useAuthEventHandler hook
 * @property getAccessToken - Function to retrieve the current access token
 * @property handleSignIn - Function to initiate the sign in flow
 * @property handleSignOut - Function to sign out the current user
 */
interface UseAuthEventHandlerReturn {
    getAccessToken: () => Promise<string>;
    handleSignIn: () => Promise<void>;
    handleSignOut: () => Promise<void>;
}

/**
 * Custom hook to handle authentication events and token management
 * @param checkUser - Function to verify the current user's authentication status
 * @param resetUserState - Function to clear user state on sign out or auth failure
 * @returns {UseAuthEventHandlerReturn}
 */
export const useAuthEventHandler = (
    checkUser: () => Promise<void>,
    resetUserState: () => void
): UseAuthEventHandlerReturn => {
    const runtimeConfig = useSelector((state: RootState) => state.config.runtimeConfig);
    const userPoolClientId = runtimeConfig?.UserPoolClientId;

    /**
     * Retrieves the current access token from the auth session
     * @returns Promise resolving to the access token string
     * @throws Error if no access token is available
     */
    const getAccessToken = useCallback(async (): Promise<string> => {
        try {
            const session = await fetchAuthSession();
            const accessToken = session.tokens?.accessToken?.toString();
            if (!accessToken) {
                throw new Error('No access token available');
            }
            return accessToken;
        } catch (error) {
            console.error('Error getting access token:', error);
            throw error;
        }
    }, []);

    /**
     * Initiates the sign in flow using redirect
     */
    const handleSignIn = async () => {
        try {
            await signInWithRedirect();
        } catch (error) {
            console.error('Error starting sign in:', error);
        }
    };

    /**
     * Signs out the current user and resets the user state
     */
    const handleSignOut = async () => {
        try {
            await signOut();
            resetUserState();
        } catch (error) {
            console.error('Error signing out:', error);
        }
    };

    /**
     * Sets up an auth event listener to handle various authentication state changes
     */
    useEffect(() => {
        const listener = Hub.listen('auth', ({ payload }) => {
            const eventHandlers: Record<string, () => void> = {
                signedIn: checkUser,
                signedOut: resetUserState,
                tokenRefresh: checkUser,
                tokenRefresh_failure: resetUserState
            };

            const handler = eventHandlers[payload.event];
            if (handler) {
                handler();
            }
        });

        return () => listener();
    }, [checkUser, resetUserState]);

    // Cross-tab auth state listener
    useEffect(() => {
        if (!userPoolClientId) return;

        const handleStorageChange = async (event: StorageEvent) => {
            const lastAuthUserKey = `CognitoIdentityServiceProvider.${userPoolClientId}.LastAuthUser`;
            const accessTokenKey = `CognitoIdentityServiceProvider.${userPoolClientId}`;

            const isSignOutEvent =
                (event.key === lastAuthUserKey && event.newValue === null) ||
                (event.key?.startsWith(accessTokenKey) &&
                    event.key?.includes('accessToken') &&
                    event.newValue === null);

            if (isSignOutEvent) {
                try {
                    // Clean up local state
                    resetUserState();
                    await signOut();
                } catch (error) {
                    console.error('Error during cross-tab signout:', error);
                }
            }
        };

        window.addEventListener('storage', handleStorageChange);
        return () => window.removeEventListener('storage', handleStorageChange);
    }, [userPoolClientId, resetUserState]);

    return { getAccessToken, handleSignIn, handleSignOut };
};
