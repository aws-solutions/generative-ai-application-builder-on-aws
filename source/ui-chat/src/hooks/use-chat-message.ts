// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { useCallback, useReducer } from 'react';
import { ChatResponse, isChatSuccessResponse, isErrorResponse, SourceDocument } from '../models';
import { ChatActionTypes, Message } from '../pages/chat/types';
import { useUser } from '../contexts/UserContext';
import { chatReducer } from '../reducers/chat-reducer';
import { END_CONVERSATION_TOKEN } from '../utils/constants';

/**
 * Interface representing the state of a chat conversation
 * @property messages - Array of chat messages
 * @property currentResponse - Current response being generated
 * @property isGenAiResponseLoading - Flag indicating if AI is generating a response
 * @property sourceDocuments - Array of source documents referenced in the conversation
 * @property conversationId - Unique identifier for the conversation
 */
export interface ChatState {
    messages: Message[];
    currentResponse: string;
    isGenAiResponseLoading: boolean;
    sourceDocuments: SourceDocument[];
    conversationId: string;
}

/**
 * Initial state for the chat
 */
const initialState: ChatState = {
    messages: [],
    currentResponse: '',
    isGenAiResponseLoading: false,
    sourceDocuments: [],
    conversationId: ''
};

/**
 * Custom hook to manage chat messages and state
 * @returns {Object} Chat state and handler functions
 */
export const useChatMessages = () => {
    const [state, dispatch] = useReducer(chatReducer, initialState);
    const { userId } = useUser();

    /**
     * Handles incoming chat responses and updates state accordingly
     * @param {ChatResponse} response - The chat response to process
     */
    const handleMessage = useCallback(
        (response: ChatResponse) => {
            try {
                if (!response) {
                    console.error('Received invalid response');
                    return;
                }

                if (isErrorResponse(response) && response.errorMessage) {
                    dispatch({
                        type: ChatActionTypes.SET_ERROR,
                        payload: response.errorMessage
                    });
                }

                if (isChatSuccessResponse(response) && response.conversationId) {
                    dispatch({ type: ChatActionTypes.SET_CONVERSATION_ID, payload: response.conversationId });
                }

                if (isChatSuccessResponse(response) && response.sourceDocument) {
                    dispatch({ type: ChatActionTypes.ADD_SOURCE_DOCUMENT, payload: response.sourceDocument });
                }

                if (isChatSuccessResponse(response) && response.rephrased_query) {
                    dispatch({ type: ChatActionTypes.ADD_REPHRASED_QUERY, payload: response.rephrased_query });
                }

                if (response.data !== undefined) {
                    if (response.data === END_CONVERSATION_TOKEN) {
                        dispatch({ type: ChatActionTypes.COMPLETE_AI_RESPONSE });
                    } else {
                        // Send only the new data token, not the concatenated response
                        dispatch({ 
                            type: ChatActionTypes.UPDATE_AI_RESPONSE, 
                            payload: { 
                                content: response.data,
                                messageId: response.messageId 
                            } 
                        });
                    }
                }
            } catch (error) {
                dispatch({
                    type: ChatActionTypes.SET_ERROR,
                    payload: 'An error occurred while processing the message.'
                });
            }
        },
        [state.currentResponse]
    );

    /**
     * Resets the chat state to initial values
     */
    const resetChat = useCallback(() => {
        dispatch({ type: ChatActionTypes.RESET_CHAT });
    }, []);

    /**
     * Updates the messages array in the chat state
     * @param {Message[]} messages - Array of messages to set
     */
    const handleSetMessages = (messages: Message[]) => {
        dispatch({ type: ChatActionTypes.SET_MESSAGES, payload: messages });
    };

    /**
     * Adds a new user message to the chat
     * @param {string} userInput - The message content from the user
     */
    const handleAddUserMessage = (userInput: string) => {
        dispatch({ type: ChatActionTypes.ADD_USER_MESSAGE, payload: { content: userInput, authorId: userId } });
    };

    return {
        messages: state.messages,
        currentResponse: state.currentResponse,
        isGenAiResponseLoading: state.isGenAiResponseLoading,
        sourceDocuments: state.sourceDocuments,
        conversationId: state.conversationId,
        handleMessage,
        addUserMessage: handleAddUserMessage,
        resetChat,
        setMessages: handleSetMessages,
        setIsGenAiResponseLoading: (loading: boolean) => {
            if (!loading) {
                dispatch({ type: ChatActionTypes.COMPLETE_AI_RESPONSE });
            }
        }
    };
};
