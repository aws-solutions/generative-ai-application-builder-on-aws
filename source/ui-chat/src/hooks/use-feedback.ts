// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { useState } from 'react';
import { useSelector } from 'react-redux';
import { useSubmitFeedbackMutation } from '../store/solutionApi';
import { getUseCaseConfigKey, getUseCaseId } from '../store/configSlice';
import { FeedbackFormData, FeedbackType } from '../pages/chat/components/input/FeedbackForm';
import { ChatBubbleMessage } from '../pages/chat/types';

export interface UseFeedbackResult {
    showFeedbackForm: boolean;
    setShowFeedbackForm: (show: boolean) => void;
    feedbackType: FeedbackType;
    setFeedbackType: (type: FeedbackType) => void;
    feedbackSubmitted: boolean;
    feedbackError: string | null;
    isSubmittingFeedback: boolean;
    handleFeedbackButtonClick: (type: FeedbackType) => void;
    handleFeedbackSubmit: (feedbackData: FeedbackFormData) => void;
}

/**
 * Custom hook to manage feedback state and submission
 * @param message The message to provide feedback for
 * @param onFeedback Callback function when feedback is submitted
 * @returns Object containing feedback state and handlers
 */
export const useFeedback = (message: ChatBubbleMessage, conversationId: string): UseFeedbackResult => {
    const [showFeedbackForm, setShowFeedbackForm] = useState(false);
    const [feedbackType, setFeedbackType] = useState<FeedbackType>('');
    const [feedbackSubmitted, setFeedbackSubmitted] = useState(false);
    const [feedbackError, setFeedbackError] = useState<string | null>(null);
    const [submitFeedback, { isLoading: isSubmittingFeedback }] = useSubmitFeedbackMutation();
    const useCaseConfigKey = useSelector(getUseCaseConfigKey);
    const useCaseId = useSelector(getUseCaseId);

    const handleFeedbackButtonClick = (type: FeedbackType) => {
        setFeedbackType(type);
        setShowFeedbackForm(true);
    };

    const handleFeedbackSubmit = (feedbackData: FeedbackFormData) => {
        if (!useCaseId) {
            console.error('Use case ID not found in runtime config');
            return;
        }
        if (!useCaseConfigKey) {
            console.error('Use case config record key not found in runtime config');
            return;
        }

        if (!message.messageId) {
            console.error('Message ID not found for feedback submission');
            setFeedbackError('Unable to submit feedback: Message ID not found');
            return;
        }
        
        const feedbackPayload = {
            useCaseRecordKey: useCaseConfigKey,
            conversationId: conversationId,
            messageId: message.messageId,
            feedback: feedbackType === 'helpful' ? 'positive' : 'negative',
            feedbackReason: feedbackData.reasons,
            comment: feedbackData.comment,
            ...(message.rephrasedQuery && { rephrasedQuery: message.rephrasedQuery }),
            ...((message.sourceDocuments && message.sourceDocuments.length > 0) && { 
                sourceDocuments: message.sourceDocuments
                    .map(doc => doc.location)
                    .filter(Boolean) // Remove any undefined or null values
                    .filter((value, index, self) => self.indexOf(value) === index) // Remove duplicates
            }),
        };

        submitFeedback({
            useCaseId,
            feedbackData: feedbackPayload
        })
            .unwrap()
            .then(() => {
                setShowFeedbackForm(false);
                setFeedbackSubmitted(true);
                setFeedbackError(null);
            })
            .catch((error) => {
                console.error('Error submitting feedback:', error);
                setShowFeedbackForm(false);
                setFeedbackSubmitted(false);
                setFeedbackError(error.data?.message || 'Failed to submit feedback. Please try again or contact a system administrator.');
            });
    };

    return {
        showFeedbackForm,
        setShowFeedbackForm,
        feedbackType,
        setFeedbackType,
        feedbackSubmitted,
        feedbackError,
        isSubmittingFeedback,
        handleFeedbackButtonClick,
        handleFeedbackSubmit
    };
};
