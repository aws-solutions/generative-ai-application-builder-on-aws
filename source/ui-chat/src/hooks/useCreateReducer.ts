/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { useMemo, useReducer } from 'react';

// Extracts property names from initial state of reducer to allow typesafe dispatch objects
export type FieldNames<T> = {
    [K in keyof T]: T[K] extends string ? K : K;
}[keyof T];

// Returns the Action Type for the dispatch object to be used for typing in things like context
export type ActionType<T> = { type: 'reset' } | { type?: 'change'; field: FieldNames<T>; value: any };

// Returns a typed dispatch and state
export const useCreateReducer = <T>({ initialState }: { initialState: T }) => {
    type Action = { type: 'reset' } | { type?: 'change'; field: FieldNames<T>; value: any };

    const reducer = (state: T, action: Action) => {
        if (!action.type) return { ...state, [action.field]: action.value };

        if (action.type === 'reset') return initialState;

        throw new Error();
    };

    const [state, dispatch] = useReducer(reducer, initialState);

    return useMemo(() => ({ state, dispatch }), [state, dispatch]);
};
