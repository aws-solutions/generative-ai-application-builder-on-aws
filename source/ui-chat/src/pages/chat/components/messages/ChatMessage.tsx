// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ChatBubbleMessage, Message } from '../../types';
import { AUTHORS, AUTHORS_CONFIG, getUserAuthor } from '../../config';
import { OutgoingMessage } from './OutgoingMessage';
import { IncomingMessage } from './IncomingMessage';

/**
 * Interface defining the props for the ChatMessage component
 * @interface ChatMessageProps
 * @property {Message} message - The message object containing content and metadata
 * @property {string} userId - Unique identifier for the user
 * @property {string} userName - Display name of the user
 * @property {Function} isUserMessage - Function to determine if a message is from the user
 * @property {Function} onFeedback - Callback function for handling message feedback
 * @property {string} [data-testid] - Optional test identifier for the component
 */
export interface ChatMessageProps {
    message: Message;
    userId: string;
    userName: string;
    isUserMessage: (authorId: string) => boolean;
    conversationId: string;
    'data-testid'?: string;
}

/**
 * Component that renders a chat message bubble with author avatar, content and actions
 * @param {ChatMessageProps} props - Component props
 * @returns {JSX.Element} Rendered chat message bubble
 */
export const ChatMessage = ({
    message,
    userId,
    userName,
    isUserMessage,
    conversationId,
    'data-testid': dataTestId
}: ChatMessageProps) => {
    const typedMessage = message as ChatBubbleMessage;
    const author =
        typedMessage.authorId === AUTHORS.ASSISTANT
            ? AUTHORS_CONFIG[AUTHORS.ASSISTANT]
            : getUserAuthor(userId, userName!);

    const shouldShowActions = typedMessage.authorId === AUTHORS.ASSISTANT && !typedMessage.avatarLoading;

    // Find the corresponding user message that came before this assistant message
    const findUserInput = () => {
        if (typedMessage.authorId === AUTHORS.ASSISTANT && typedMessage.userInput) {
            return typedMessage.userInput;
        }
        return undefined;
    };

    // Ensure the message ID is passed to the IncomingMessage component
    return isUserMessage(typedMessage.authorId) ? (
        <OutgoingMessage message={typedMessage} author={author} data-testid={`outgoing-${dataTestId}`} />
    ) : (
        <IncomingMessage
            message={{
                ...typedMessage,
                userInput: findUserInput()
            }}
            author={author}
            showActions={shouldShowActions}
            conversationId={conversationId}
            data-testid={`incoming-${dataTestId}`}
        />
    );
};
