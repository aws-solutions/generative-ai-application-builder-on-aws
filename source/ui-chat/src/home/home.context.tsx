/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Dispatch, createContext } from 'react';

import { ActionType } from '../hooks/useCreateReducer';

import { Conversation } from '../types/chat';
import { KeyValuePair } from '../types/misc';

import { HomeInitialState } from './home.state';

export interface HomeContextProps {
    state: HomeInitialState;
    dispatch: Dispatch<ActionType<HomeInitialState>>;
    handleUpdateConversation: (conversation: Conversation, data: KeyValuePair) => void;
}

const HomeContext = createContext<HomeContextProps>(undefined!);

export default HomeContext;
