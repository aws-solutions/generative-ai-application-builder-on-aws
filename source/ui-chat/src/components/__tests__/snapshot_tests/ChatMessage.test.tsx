/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 **********************************************************************************************************************/

import { Dispatch } from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import HomeContext from '../../../home/home.context';
import { HomeInitialState } from '../../../home/home.state';
import { ActionType } from '../../../hooks/useCreateReducer';
import { ChatMessage } from '../../ChatMessage';

jest.mock('@cloudscape-design/components', () => {
    const Components = jest.genMockFromModule('@cloudscape-design/components') as any;
    for (const componentName of Object.keys(Components)) {
        Components[componentName] = componentName;
    }
    return Components;
});

jest.mock('react-markdown', () => (props) => {
    return <>{props.children}</>;
});

jest.mock('remark-gfm', () => (props) => {
    return <>{props.children}</>;
});

describe('Chat', () => {
    const contextValue = {
        dispatch: jest.fn() as Dispatch<ActionType<HomeInitialState>>,
        state: {
            loading: false,
            messageIsStreaming: false,
            selectedConversation: {
                id: '',
                name: 'New Conversation',
                messages: []
            },
            promptTemplate: '',
            defaultPromptTemplate: ''
        },
        handleUpdateConversation: jest.fn()
    };

    test('Snapshot test for chat message from assistant', async () => {
        const tree = renderer
            .create(
                <>
                    <HomeContext.Provider
                        value={{
                            ...contextValue
                        }}
                    >
                        <ChatMessage message={{ role: 'assistant', content: 'hello' }} messageIndex={1} />
                    </HomeContext.Provider>
                </>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('Snapshot test for chat message from user', async () => {
        const tree = renderer
            .create(
                <>
                    <HomeContext.Provider
                        value={{
                            ...contextValue
                        }}
                    >
                        <ChatMessage message={{ role: 'user', content: 'hello' }} messageIndex={1} />
                    </HomeContext.Provider>
                </>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
