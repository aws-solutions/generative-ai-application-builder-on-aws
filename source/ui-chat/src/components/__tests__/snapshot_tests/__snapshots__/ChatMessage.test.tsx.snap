// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chat > Snapshot test for chat message from assistant 1`] = `
<div
  className="bg-gray-100"
  style={
    {
      "overflowWrap": "anywhere",
    }
  }
>
  <div
    className=" m-auto flex md:gap-6 md:py-6 xl:max-w-2xl"
  >
    <div
      className="text-right font-bold "
    >
      <span
        className="awsui_icon_h11ix_ayzd6_101 awsui_size-normal-mapped-height_h11ix_ayzd6_157 awsui_size-normal_h11ix_ayzd6_153 awsui_variant-normal_h11ix_ayzd6_229"
        style={{}}
      >
        <svg
          aria-hidden="true"
          focusable="false"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            className="stroke-linejoin-round stroke-linecap-round"
            d="M2 12.286h5.143L8.857 14l1.714-1.714H14V2H2v10.286z"
          />
          <path
            className="stroke-linecap-round"
            d="M4.99 7H5v.01h-.01zM7.99 7H8v.01h-.01zM10.99 7H11v.01h-.01z"
          />
        </svg>
      </span>
    </div>
    <div
      className="prose mt-[-2px] w-full"
    >
      <div
        className="flex "
      >
        <div
          className="flex-1"
        >
          <div
            className="prose dark:prose-invert flex-1"
          >
            <p>
              hello
            </p>
          </div>
        </div>
        <div
          className="copy-button-div md:-mr-8"
        >
          <div
            className="copy-button-div-nested"
          >
            <button
              className="awsui_button_vjswe_15u39_105 awsui_variant-icon_vjswe_15u39_199 awsui_button-no-text_vjswe_15u39_1085"
              data-analytics-funnel-value="button:r0:"
              data-testid="copy-msg-button"
              disabled={false}
              onClick={[Function]}
              type="submit"
            >
              <span
                className="awsui_icon_vjswe_15u39_1111 awsui_icon-left_vjswe_15u39_1111 awsui_icon_h11ix_ayzd6_101 awsui_size-normal-mapped-height_h11ix_ayzd6_157 awsui_size-normal_h11ix_ayzd6_153 awsui_variant-normal_h11ix_ayzd6_229"
                style={{}}
              >
                <svg
                  aria-hidden="true"
                  focusable="false"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    className="stroke-linejoin-round"
                    d="M2 5h9v9H2z"
                  />
                  <path
                    className="stroke-linejoin-round"
                    d="M5 5V2h9v9h-3"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Chat > Snapshot test for chat message from user 1`] = `
<div
  className=""
  style={
    {
      "overflowWrap": "anywhere",
    }
  }
>
  <div
    className=" m-auto flex md:gap-6 md:py-6 xl:max-w-2xl"
  >
    <div
      className="prose mt-[-2px] w-full"
    >
      <div
        className="w-full"
      >
        <p>
          hello
        </p>
      </div>
    </div>
    <div
      className="min-w-[40px] text-right"
    >
      <span
        className="awsui_icon_h11ix_ayzd6_101 awsui_size-normal-mapped-height_h11ix_ayzd6_157 awsui_size-normal_h11ix_ayzd6_153 awsui_variant-normal_h11ix_ayzd6_229"
        style={{}}
      >
        <svg
          aria-hidden="true"
          focusable="false"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            className="stroke-linejoin-round"
            d="M4.61 4.39C4.61 2.52 6.13 1 8 1s3.39 1.52 3.39 3.39S9.87 7.78 8 7.78 4.61 6.26 4.61 4.39M2.05 14.08h11.91c0-2.25-1.17-4.23-2.9-5.3A5.77 5.77 0 0 0 8 7.91c-3.29 0-5.95 2.76-5.95 6.17z"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
